client.go

```diff
@@ -0,0 +1,593 @@
+	// milter using a newer protocol as long as masks negotiated are meaningful.
+	if milterVersion < protocolVersion {
+		return fmt.Errorf("milter: negotiate: unsupported protocol version: %v", milterVersion)
+	}
+
+	// AND it with our mask in case milter does not do that.
+	milterActionMask := binary.BigEndian.Uint32(msg.Data[4:])
+	s.actionMask = actionMask & OptAction(milterActionMask)
+	milterProtoMask := binary.BigEndian.Uint32(msg.Data[8:])
+	s.protoMask = protoMask & OptProtocol(milterProtoMask)
+
+	s.needAbort = true
+
+	return nil
+}
+
+// ProtocolOption checks whether the option is set in negotiated options, that
+// is, requested by both sides.
+func (s *ClientSession) ProtocolOption(opt OptProtocol) bool {
+	return s.protoMask&opt != 0
+}
+
+// ActionOption checks whether the option is set in negotiated options, that
+// is, requested by both sides.
+func (s *ClientSession) ActionOption(opt OptAction) bool {
+	return s.actionMask&opt != 0
+}
+
+func (s *ClientSession) Macros(code Code, kv ...string) error {
+	// Note: kv is ...string with the expectation that the list of macro names
+	// will be static and not dynamically constructed.
+
+	msg := &Message{
+		Code: byte(CodeMacro),
+		Data: []byte{byte(code)},
+	}
+	for _, str := range kv {
+		msg.Data = appendCString(msg.Data, str)
+	}
+
+	if err := writePacket(s.conn, msg, s.writeTimeout); err != nil {
+		return fmt.Errorf("milter: macros: %w", err)
+	}
+
+	return nil
+}
+
+func appendUint16(dest []byte, val uint16) []byte {
+	dest = append(dest, 0x00, 0x00)
+	binary.BigEndian.PutUint16(dest[len(dest)-2:], val)
+	return dest
+}
+
+type Action struct {
+	Code ActionCode
+
+	// SMTP code if Code == ActReplyCode.
+	SMTPCode int
+	// Reply text if Code == ActReplyCode.
+	SMTPText string
+}
+
+func (s *ClientSession) readAction() (*Action, error) {
+	for {
+		msg, err := readPacket(s.conn, s.readTimeout)
+		if err != nil {
+			return nil, fmt.Errorf("action read: %w", err)
+		}
+		if msg.Code == 'p' /* progress */ {
+			continue
+		}
+		if ActionCode(msg.Code) != ActContinue {
+			s.needAbort = false
+		}
+
+		return parseAction(msg)
+	}
+}
+
+func parseAction(msg *Message) (*Action, error) {
+	act := &Action{
+		Code: ActionCode(msg.Code),
+	}
+	var err error
+
+	switch ActionCode(msg.Code) {
+	case ActAccept, ActContinue, ActDiscard, ActReject, ActTempFail:
+	case ActReplyCode:
+		if len(msg.Data) <= 4 {
+			return nil, fmt.Errorf("action read: unexpected data length: %v", len(msg.Data))
+		}
+		act.SMTPCode, err = strconv.Atoi(string(msg.Data[:3]))
+		if err != nil {
+			return nil, fmt.Errorf("action read: malformed SMTP code: %v", msg.Data[:3])
+		}
+		// There is 0x20 (' ') in between.
+		act.SMTPText = readCString(msg.Data[4:])
+	default:
+		return nil, fmt.Errorf("action read: unexpected code: %v", msg.Code)
+	}
+
+	return act, nil
+}
+
+// Helo sends the connection information to the milter.
+//
+// It should be called once per milter session (from NewSession to Close).
```

Comment:
`NewSession` has been renamed to `Session` it seems

Reply:
Fixed in 563ba50.